src/main/java/org/mule/service/oauth/internal/AbstractOAuthDancer.java

```diff
@@ -138,6 +164,95 @@ protected String handleClientCredentials(final Map<String, String> formData) {
     return null;
   }
 
+  /**
+   * Method for refreshing tokens in a thread-safe manner across nodes of a cluster.
+   */
+  protected CompletableFuture<Void> doRefreshToken(Supplier<ResourceOwnerOAuthContext> oauthContextSupplier,
+                                                   Function<ResourceOwnerOAuthContext, CompletableFuture<Void>> tokenRefreshRequester) {
+    ResourceOwnerOAuthContext oauthContext = oauthContextSupplier.get();
+
+    final Lock lock = oauthContext.getRefreshUserOAuthContextLock(name, getLockProvider());
+
+    // If the context was just created, initialize it.
+    if (oauthContext.getDancerState() == NO_TOKEN) {
+      if (lock.tryLock()) {
+        try {
+          oauthContext = oauthContextSupplier.get();
+          if (oauthContext.getDancerState() == HAS_TOKEN) {
+            // Some other thread/node completed the refresh before lock was acquired here. Very quickly and quite improbable, but
+            // possible.
+            return completedFuture(null);
+          } else if (oauthContext.getDancerState() == REFRESHING_TOKEN) {
+            return pollForRefreshComplete(oauthContextSupplier);
+          } else if (oauthContext.getDancerState() == NO_TOKEN) {
+            ((DefaultResourceOwnerOAuthContext) oauthContext).setDancerState(REFRESHING_TOKEN);
+            updateResourceOwnerOAuthContext(oauthContext);
+
+            return tokenRefreshRequester.apply(oauthContext);
+          }
+        } finally {
+          lock.unlock();
+        }
+      } else {
+        return pollForRefreshComplete(oauthContextSupplier);
+      }
+    }
+
+    // If there is a previous token, refresh it
+    if (oauthContext.getDancerState() == HAS_TOKEN) {
+      final String accessToken = oauthContext.getAccessToken();
+      lock.lock();
```

Comment:
use org.mule.runtime.core.internal.util.ConcurrencyUtils#withLock(java.util.concurrent.locks.Lock, org.mule.runtime.core.api.util.func.CheckedSupplier<T>). Applies to the entire PR

Reply:
as the package name suggests, `ConcurrencyUtils` is not api and cannot be used within a service